File: firmware/omsk_grain/src/synth/audio_engine.cpp
#include "audio_engine.h"
#include "fx.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include "../sw_config.h"
#if __has_include("../tables/audio_data.h")
#include "../tables/audio_data.h"
#else
#include "../tables/audio_data_stub.h"
#endif
#include "../tables/eg_lut.h"
#include "fast_math.h"

// ------------------------------------------------------------------ //
//  LOOKUP TABLES
// ------------------------------------------------------------------ //

static float SINE_LUT[512];
static float POW2_LUT[512]; // for 2^x, x in [0, 10]
static float NOTE_FREQ_MULT_LUT[128];
static float SEMI_PITCH_LUT[49]; // -24 to +24
static float OCT_PITCH_LUT[7];   // -3 to +3
static bool luts_init = false;

static void init_engine_luts() {
    if (luts_init) return;
    for (int i = 0; i < 512; i++) {
        SINE_LUT[i] = sinf(i * 2.0f * M_PI / 512.0f);
        POW2_LUT[i] = powf(2.0f, (i / 511.0f) * 10.0f); // Map 0..1 to 2^0..2^10
    }
    for (int i = 0; i < 128; i++) {
        NOTE_FREQ_MULT_LUT[i] = powf(2.0f, (i - 60.0f) / 12.0f);
    }
    for (int i = 0; i < 49; i++) {
        SEMI_PITCH_LUT[i] = powf(2.0f, (i - 24) / 12.0f);
    }
    for (int i = 0; i < 7; i++) {
        OCT_PITCH_LUT[i] = powf(2.0f, i - 3);
    }
    luts_init = true;
}

static inline float get_sine_lut(float phase) {
    while (phase < 0) phase += 1.0f;
    while (phase >= 1.0f) phase -= 1.0f;
    int idx = (int)(phase * 511.99f);
    return SINE_LUT[idx & 511];
}

static inline float get_exp_freq(float ratio) {
    if (ratio < 0.0f) ratio = 0.0f;
    if (ratio > 1.0f) ratio = 1.0f;
    int idx = (int)(ratio * 511.99f);
    return 20.0f * POW2_LUT[idx & 511];
}

static float get_curved_env(float phase, float curve) {
    if (phase < 0.0f) phase = 0.0f;
    if (phase > 1.0f) phase = 1.0f;
    int c_idx = (int)((curve + 1.0f) * 10.0f + 0.5f);
    if (c_idx < 0) c_idx = 0;
    if (c_idx >= NUM_EG_CURVES) c_idx = NUM_EG_CURVES - 1;
    int t_idx = (int)(phase * (EG_LUT_SIZE - 1.0f));
    return EG_LUT[c_idx][t_idx];
}

// ------------------------------------------------------------------ //
//  ENGINE
// ------------------------------------------------------------------ //

GranularEngine::GranularEngine(int vid) : voice_id(vid) {
    init_engine_luts();
    memset(&p, 0, sizeof(p));
    memset(mod_src, 0, sizeof(mod_src));
    memset(mod_amt, 0, sizeof(mod_amt));
    
    p.pos = 0.5f;
    p.size = 0.1f;
    p.dens = 20.0f;
    p.pitch = 1.0f;
    p.max_grains = 32;
    p.grain_amp = 0.8f;
    p.keytrack = 1.0f; // Enable pitch keytracking by default (100%)
    p.cutoff = 20000.0f;
    p.res = 0.7f;
    p.filt_type = 3; // Off
    p.atk = 0.01f;
    p.rel = 0.3f;
    p.vol = 0.7f;
    p.master_vol = 1.0f;
    p.midi_ch = 1;
    p.midi_mode = 0;
    p.viz_scale = 1.0f;
    p.fx_bc = 16.0f;
    p.fx_ds = 1.0f;
    
    playback_pos = 0;
    next_grain_time = 0;
    is_triggered = false;
    current_note = -1;
    last_note = 60; // Default to reference C4 note (1.0x ratio)
    pitch_bend_semitones = 0.0f;
    modwheel_val = 0.0f;
    aftertouch_val = 0.0f;
    sustain_active = false;
    is_sustained = false;
    lin_env = 0;

    master_env = 0;
    lfo_phases[0] = 0;
    lfo_phases[1] = 0;
    
    fx_processor.reset();
}

void GranularEngine::set_trigger(bool triggered) {
    is_triggered = triggered;
}

float GranularEngine::get_modulated_param(int param_id, float base_val, float lo, float hi) {
    if (param_id >= 64) return base_val;
    uint8_t src = mod_src[param_id];
    if (src == 0) return base_val;
    float amt = mod_amt[param_id];
    float mod = get_mod_val(src);
    
    float mod_val;
    if (src == 4) { // EG (bipolar -1..1)
        // get_mod_val returns 0..1, map to -1..1
        mod_val = (mod - 0.5f) * 2.0f;
    } else { // Jitter, LFOs (bipolar -1..1)
        mod_val = mod * 2.0f - 1.0f;
    }
    
    if (param_id == 29) { // PARAM_CUTOFF
        // Logarithmic modulation
        float ratio = log2f(base_val / 20.0f) / 10.0f; // 10 octaves
        ratio += amt * mod_val;
        return get_exp_freq(ratio);
    }
    
    float range = hi - lo;
    float result = base_val + amt * mod_val * range;
    
    if (result < lo) result = lo;
    if (result > hi) result = hi;
    return result;
}

float GranularEngine::get_mod_val(int mod_src_id) {
    switch(mod_src_id) {
        case 1: return (float)(rand() % 1000) / 1000.0f; // Jitter
        case 2: return get_lfo_val(0) * 0.5f + 0.5f;     // LFO1
        case 3: return get_lfo_val(1) * 0.5f + 0.5f;     // LFO2
        case 4: { // EG
            if (master_env >= 0.999f || is_triggered) {
                return get_curved_env(master_env, p.atk_curve);
            } else {
                float t_norm = 1.0f - master_env;
                return 1.0f - get_curved_env(t_norm, -p.rel_curve);
            }
        }
        case 5: return modwheel_val;                     // Modwheel
        case 6: return aftertouch_val;                   // Aftertouch
        default: return 0;

    }
}

bool GranularEngine::is_lfo_used(int lfo_idx) {
    uint8_t target_src = (lfo_idx == 0) ? 2 : 3;
    for (int i = 0; i < 64; i++) {
        if (mod_src[i] == target_src) return true;
    }
    return false;
}

void GranularEngine::process(int frames, float* out_l, float* out_r, const float* fm_signal, const float* rm_signal, float rm_amt) {
    // 1. Optimization: Voice skipping
    if (!is_triggered && master_env <= 0.001f) {
        for (int g = 0; g < MAX_GRAINS_PER_VOICE; g++) {
            active_grains[g].active = false;
        }
        memset(out_l, 0, frames * sizeof(float));
        memset(out_r, 0, frames * sizeof(float));
        memset(last_buffer, 0, frames * sizeof(float));
        return;
    }

    // Keep track of the last active MIDI note
    if (current_note >= 0) {
        last_note = current_note;
    }

    float dt = (float)frames / SAMPLE_RATE;
    
    // 2. Optimization: LFO phase update
    if (is_lfo_used(0)) lfo_phases[0] = fmodf(lfo_phases[0] + p.lfo1_rate * dt, 1.0f);
    if (is_lfo_used(1)) lfo_phases[1] = fmodf(lfo_phases[1] + p.lfo2_rate * dt, 1.0f);

    float target = is_triggered ? 1.0f : 0.0f;
    if (target > master_env) {
        master_env += dt / (p.atk + 0.001f);
        if (master_env > target) master_env = target;
    } else {
        master_env -= dt / (p.rel + 0.001f);
        if (master_env < target) master_env = target;
    }

    float mod_cutoff = get_modulated_param(29, p.cutoff, 20.0f, 20000.0f);
    
    // Keytrack logic
    int note_for_track = (current_note >= 0) ? current_note : last_note;
    if (p.filt_key > 0.01f && note_for_track >= 0) {
        float freq_mult = 1.0f;
        if (note_for_track < 128) freq_mult = NOTE_FREQ_MULT_LUT[note_for_track];
        mod_cutoff *= (1.0f - p.filt_key) + (freq_mult * p.filt_key);
        if (mod_cutoff > 20000.0f) mod_cutoff = 20000.0f;
    }

    float mod_res_q  = get_modulated_param(30, p.res, 0.5f, 13.0f);
    float mod_res    = 1.0f - 0.5f / mod_res_q;
    if (mod_res < 0.0f) mod_res = 0.0f;
    if (mod_res > 0.99f) mod_res = 0.99f;
    float mod_vol    = get_modulated_param(41, p.vol, 0.0f, 1.0f);

    float curved_env = 0.0f;
    if (master_env > 0.001f) {
        if (master_env >= 0.999f || is_triggered) {
            curved_env = get_curved_env(master_env, p.atk_curve);
        } else {
            float t_norm = 1.0f - master_env;
            curved_env = 1.0f - get_curved_env(t_norm, -p.rel_curve);
        }
    }
    float gain = mod_vol * curved_env * p.master_vol;

    if (gain <= 0.001f) {
        if (is_triggered) {
            playback_pos = fmodf(playback_pos + p.scan * dt, 1.0f);
            if (playback_pos < 0.0f) playback_pos += 1.0f;
        }
        memset(out_l, 0, frames * sizeof(float));
        memset(out_r, 0, frames * sizeof(float));
        memset(last_buffer, 0, frames * sizeof(float));
        return;
    }

    if (is_triggered) {
        playback_pos = fmodf(playback_pos + p.scan * dt, 1.0f);
        if (playback_pos < 0.0f) playback_pos += 1.0f;
    }
    next_grain_time -= frames;
    if (next_grain_time <= 0 && num_active_grains() < p.max_grains) {
        spawn_grain();
        float dens = get_modulated_param(2, p.dens, 0.1f, 100.0f);
        next_grain_time = SAMPLE_RATE / dens;
    }

    for (int i = 0; i < frames; i++) {
        float l = 0, r = 0;
        
        float fm = 0;
        if (fm_signal) fm = fm_signal[i];

        for (int g = 0; g < MAX_GRAINS_PER_VOICE; g++) {
            if (active_grains[g].active) {
                active_grains[g].process(fm, &l, &r);
            }
        }
        
        // 3. RM Bypass Logic
        if (rm_signal && rm_amt > 0.001f) {
            float rm = rm_signal[i];
            float rm_mix = (1.0f - rm_amt) + (rm * rm_amt * 5.0f); // Multiply the signal by RM amount
            l = l * rm_mix;
            r = r * rm_mix;
        }

        if (p.filt_type < 3) {
            l = filter_l.process(l, mod_cutoff, mod_res, p.filt_type, SAMPLE_RATE);
            r = filter_r.process(r, mod_cutoff, mod_res, p.filt_type, SAMPLE_RATE);
        }

        // FX moved to global Synth::process

        out_l[i] = l * gain;
        out_r[i] = r * gain;
        last_buffer[i] = (out_l[i] + out_r[i]) * 0.5f;
    }
}

void GranularEngine::spawn_grain() {
    for (int i = 0; i < MAX_GRAINS_PER_VOICE; i++) {
        if (!active_grains[i].active) {
            float mod_pos   = get_modulated_param(0, p.pos, 0.0f, 1.0f);
            float mod_size  = get_modulated_param(1, p.size, 0.01f, 2.0f);
            // Apply JIT pitch modulation as integer semitone/octave steps depending on pitch mode
            float effective_pitch = p.pitch;
            if (p.pitch_mode == 1) { // SEMI (semitone steps)
                int semi_idx = (int)p.pitch + 24;
                // Clamp base index
                if (semi_idx < 0) semi_idx = 0;
                if (semi_idx > 48) semi_idx = 48;
                // Get JIT value (0..1)
                float jit_val = get_mod_val(1);
                // Convert to bipolar -1..1 and scale to integer steps
                int jit_steps = (int)roundf((jit_val * 2.0f - 1.0f) * 24.0f); // +/-24 semitones max
                int new_idx = semi_idx + jit_steps;
                if (new_idx < 0) new_idx = 0;
                if (new_idx > 48) new_idx = 48;
                effective_pitch = SEMI_PITCH_LUT[new_idx];
            } else if (p.pitch_mode == 2) { // OCT (octave steps)
                int oct_idx = (int)p.pitch + 3;
                if (oct_idx < 0) oct_idx = 0;
                if (oct_idx > 6) oct_idx = 6;
                float jit_val = get_mod_val(1);
                int jit_steps = (int)roundf((jit_val * 2.0f - 1.0f) * 3.0f); // +/-3 octaves max
                int new_idx = oct_idx + jit_steps;
                if (new_idx < 0) new_idx = 0;
                if (new_idx > 6) new_idx = 6;
                effective_pitch = OCT_PITCH_LUT[new_idx];
            } else {
                // SPEED mode – continuous pitch, apply normal modulation later
                effective_pitch = p.pitch;
            }
            // Apply other pitch modulation sources (e.g., EG, LFO) via get_modulated_param as before
            float mod_pitch = get_modulated_param(3, effective_pitch, 0.01f, 10.0f);
            
            // Apply pitch keytracking based on the MIDI note.
            // Reference note is C4 (note 60).
            int note_for_pitch = (current_note >= 0) ? current_note : last_note;
            if (p.keytrack > 0.001f && note_for_pitch >= 0) {
                float keytrack_mult = fast_exp2(p.keytrack * (note_for_pitch - 60.0f) / 12.0f);
                mod_pitch *= keytrack_mult;
            }

            // Apply Pitch Bend
            if (pitch_bend_semitones != 0.0f) {
                mod_pitch *= fast_exp2(pitch_bend_semitones / 12.0f);
            }

            
            float start_idx_ratio = fmodf(mod_pos + playback_pos, 1.0f);
            if (start_idx_ratio < 0.0f) start_idx_ratio += 1.0f;
            float start_idx = start_idx_ratio * raw_len(p.sample_idx);
            float length_frames = mod_size * SAMPLE_RATE;
            bool reverse = (float)(rand() % 100) < (p.direction * 100.0f);
            float rand_val = (float)rand() / (float)RAND_MAX;
            float grain_pan = 0.5f + (rand_val - 0.5f) * p.spread;
            
            active_grains[i].init(p.sample_idx, start_idx, length_frames, mod_pitch, grain_pan, reverse, (int)p.shape, p.grain_amp);
            break;
        }
    }
}

float GranularEngine::get_lfo_val(int lfo_idx) {
    float phase = lfo_phases[lfo_idx];
    float wave = (lfo_idx == 0) ? p.lfo1_wave : p.lfo2_wave;
    
    switch((int)wave) {
        case 0: return get_sine_lut(phase);
        case 1: return (phase < 0.5f) ? (phase * 4.0f - 1.0f) : (3.0f - phase * 4.0f); // Triangle
        case 2: return phase * 2.0f - 1.0f; // Saw
        case 3: return 1.0f - phase * 2.0f; // RSAW (Reverse Saw)
        case 4: return (float)(rand() % 2000 - 1000) / 1000.0f; // S&H (approx)
        default: return 0;
    }
}

int GranularEngine::num_active_grains() {
    int count = 0;
    for (int i = 0; i < MAX_GRAINS_PER_VOICE; i++) {
        if (active_grains[i].active) count++;
    }
    return count;
}

float GranularEngine::grain_pos_ratio(int grain_idx) {
    if (grain_idx >= MAX_GRAINS_PER_VOICE) return 0.0f;
    const Grain& g = active_grains[grain_idx];
    if (!g.active) return 0.0f;
    
    uint32_t data_len = raw_len(g.sample_id);
    if (data_len == 0) return 0.0f;
    
    return g.current_idx / (float)data_len;
}

float GranularEngine::grain_pitch(int grain_idx) {
    if (grain_idx >= MAX_GRAINS_PER_VOICE) return 1.0f;
    return active_grains[grain_idx].pitch;
}

float GranularEngine::grain_pan(int grain_idx) {
    if (grain_idx >= MAX_GRAINS_PER_VOICE) return 0.5f;
    return active_grains[grain_idx].pan;
}

File: desktop/grain_simulator/audio_engine.py
import os
import random
import numpy as np
import soundfile as sf
from constants import SAMPLE_RATE, OLED_WIDTH, generate_window

class StateVariableFilter:
    """Standard SVF filter for resonant processing."""
    def __init__(self):
        self.ic1eq = 0.0
        self.ic2eq = 0.0

    def process(self, v0, cutoff, res, mode_idx):
        if mode_idx == 3: return v0 
        cutoff = max(20, min(SAMPLE_RATE // 2.1, cutoff))
        res = max(0.01, min(0.99, res))
        g = np.tan(np.pi * cutoff / SAMPLE_RATE)
        k = 2.0 - 2.0 * res 
        a1 = 1.0 / (1.0 + g * (g + k))
        a2 = g * a1
        a3 = g * a2
        v3 = v0 - self.ic2eq
        v1 = a1 * self.ic1eq + a2 * v3
        v2 = self.ic2eq + a2 * self.ic1eq + a3 * v3
        self.ic1eq = 2.0 * v1 - self.ic1eq
        self.ic2eq = 2.0 * v2 - self.ic2eq
        if mode_idx == 0: return v2 
        if mode_idx == 1: return v0 - k * v1 - v2 
        if mode_idx == 2: return v1 
        return v0

class Grain:
    """Single grain instance handling playback logic."""
    def __init__(self, sample_data, start_idx, length, pitch, pan, reverse, window, amp):
        self.sample_data = sample_data
        self.start_idx = start_idx
        self.length = int(length)
        self.pitch = max(0.001, pitch)
        self.pan = pan 
        self.reverse = reverse
        self.amp = amp
        self.window = window
        self.current_frame = 0.0
        self.active = True

    def get_current_sample_idx(self):
        rel_pos = self.current_frame * self.pitch
        idx = (self.start_idx - rel_pos if self.reverse else self.start_idx + rel_pos)
        return int(idx) % len(self.sample_data)

    def process(self, frames, fm_signal=None):
        if not self.active or self.length <= 1: return np.zeros((frames, 2))
        out = np.zeros((frames, 2))
        data_len = len(self.sample_data)
        
        for i in range(frames):
            if self.current_frame >= self.length - 1:
                self.active = False
                break
            
            # Effective pitch taking FM into account
            shift = fm_signal[i] if fm_signal is not None else 0
            effective_pitch = max(0.001, self.pitch + shift)
                
            rel_pos = self.current_frame * effective_pitch
            idx = (self.start_idx - rel_pos if self.reverse else self.start_idx + rel_pos)
            
            # --- Linear interpolation for FM ---
            idx_low = int(idx) % data_len
            idx_high = (idx_low + 1) % data_len
            frac = idx - int(idx)
            
            sample = (1.0 - frac) * self.sample_data[idx_low] + frac * self.sample_data[idx_high]
            # ------------------------------------
            
            win_idx = int((self.current_frame / self.length) * (len(self.window)-1))
            val = sample * self.window[win_idx] * self.amp
            
            out[i, 0] = val * (1.0 - self.pan)
            out[i, 1] = val * self.pan
            self.current_frame += 1
            
        return out

class GranularEngine:
    """Core audio engine for a single granular voice."""
    SHAPES = ["Hanning", "Tri", "Rect", "Blackman"]
    FILTERS = ["LP", "HP", "BP", "Off"]

    def __init__(self, voice_id):
        self.voice_id = voice_id
        self.sample_data = np.zeros(SAMPLE_RATE)
        self.active_grains = []
        self.next_grain_time = 0
        self.current_folder = ""
        self.file_list = []
        self.current_filename = "EMPTY"
        
        self.last_buffer = np.zeros(512) 
        self.all_engines = [] 
        self.MOD_ROUTES = [] 
        
        self.params = {
            "pos": [0.5, 0, 1, 0.001, "%", None, 0],
            "size": [0.1, 0.01, 2.0, 0.001, "s", None, 0],
            "dens": [20.0, 1.0, 150.0, 0.1, "Hz", None, 0],
            "pitch": [1.0, 0.1, 4.0, 0.01, "x", None, 0],
            "sample_idx": [0, 0, 127, 1, "#", None, 0],
            "max_grains": [32, 1, 64, 1, "", None, 0],
            "grain_amp": [0.8, 0, 1, 0.01, "", None, 0],
            "keytrack": [0, 0, 1, 0.01, "", None, 0],
            "scan": [0.0, -2.0, 2.0, 0.01, "", None, 0],
            "direction": [0.0, 0, 1, 1, "", None, 0],
            "spread": [0.5, 0, 1, 0.01, "", None, 0],
            "shape": [0, 0, 3, 1, "", None, 0],
            "cutoff": [10000, 20, 20000, 1, "Hz", None, 0],
            "res": [0.1, 0, 0.99, 0.01, "", None, 0],
            "filt_type": [3, 0, 3, 1, "", None, 0],
            "filt_key": [0, 0, 1, 0.01, "", None, 0],
            "atk": [0.01, 0.001, 5.0, 0.001, "s", None, 0],
            "atk_curve": [0.0, -1.0, 1.0, 0.1, "", None, 0],
            "rel": [0.3, 0.001, 5.0, 0.001, "s", None, 0],
            "rel_curve": [0.0, -1.0, 1.0, 0.1, "", None, 0],
            "lfo1_rate": [1.0, 0.1, 50.0, 0.1, "Hz", None, 0],
            "lfo1_wave": [0, 0, 3, 1, "", None, 0],
            "lfo1_phase": [0, 0, 1, 0.01, "", None, 0],
            "lfo2_rate": [1.0, 0.1, 50.0, 0.1, "Hz", None, 0],
            "lfo2_wave": [0, 0, 3, 1, "", None, 0],
            "lfo2_phase": [0, 0, 1, 0.01, "", None, 0],
            "vol": [0.7, 0, 1, 0.01, "", None, 0],
            "mod1_src": [0, 0, 24, 1, "", None, 0], 
            "mod1_amt": [0.0, 0, 1, 0.01, "", None, 0],
            "mod2_src": [0, 0, 24, 1, "", None, 0],
            "mod2_amt": [0.0, 0, 1, 0.01, "", None, 0],
            "fx_wf": [0.0, 0, 1, 0.01, "", None, 0],
            "fx_ds": [1.0, 1, 80, 1, "", None, 0],
            "fx_bc": [16.0, 1, 16, 1, "", None, 0],
            "fx_mix": [0.0, 0, 1, 0.01, "", None, 0],
            "viz_scale": [1.0, 0.1, 5.0, 0.1, "", None, 0],
            "midi_mode": [0, 0, 6, 1, "", None, 0],
            "midi_ch": [1, 1, 16, 1, "", None, 0],
            "master_vol": [1.0, 0, 1, 0.01, "", None, 0],
            "pitch_mode": [0, 0, 2, 1, "", None, 0]
        }

        self.filter_l, self.filter_r = StateVariableFilter(), StateVariableFilter()
        self.playback_pos, self.is_triggered, self.master_env, self.curved_env = 0.0, False, 0.0, 0.0
        self.lfo_phases = [0.0, 0.0]
        self.env_pos = []
        self.window_lut = {name: generate_window(name, 1024) for name in self.SHAPES}

    def trigger_on(self):
        self.is_triggered = True

    def trigger_off(self):
        self.is_triggered = False

    def load_sample(self):
        if not self.file_list: return
        idx = int(self.params["sample_idx"][0]) % len(self.file_list)
        path = os.path.join(self.current_folder, self.file_list[idx])
        try:
            data, sr = sf.read(path)
            if len(data.shape) > 1: data = data[:, 0]
            mask = np.abs(data) > 0.02
            if np.any(mask):
                indices = np.where(mask)[0]
                data = data[indices[0]:indices[-1]]
            peak = np.max(np.abs(data))
            if peak > 0: data = data / peak
            self.sample_data = data.astype(np.float32)
            self.current_filename = self.file_list[idx]
            step = max(1, len(self.sample_data) // OLED_WIDTH)
            self.env_pos = [np.max(np.abs(self.sample_data[i:i+step])) for i in range(0, len(self.sample_data), step)][:OLED_WIDTH]
        except Exception: 
            self.current_filename = "ERROR"

    def get_mod_val(self, mod_name):
        if mod_name == "Jit": return random.uniform(-1, 1)
        if mod_name == "LFO1": return np.sin(2 * np.pi * self.lfo_phases[0])
        if mod_name == "LFO2": return np.sin(2 * np.pi * self.lfo_phases[1])
        if mod_name == "EG": return self.curved_env
        return 0

    def process_audio(self, frames):
        step = frames / SAMPLE_RATE
        self.lfo_phases[0] = (self.lfo_phases[0] + self.params["lfo1_rate"][0] * step) % 1.0
        self.lfo_phases[1] = (self.lfo_phases[1] + self.params["lfo2_rate"][0] * step) % 1.0

        out = np.zeros((frames, 2))
        if len(self.sample_data) < 100: 
            self.last_buffer = np.zeros(frames)
            return out
        
        target = 1.0 if self.is_triggered else 0.0
        if target > self.master_env: 
            self.master_env = min(target, self.master_env + step / max(0.001, self.params["atk"][0]))
        else: 
            self.master_env = max(target, self.master_env - step / max(0.001, self.params["rel"][0]))
        
        if self.master_env <= 0 and not self.is_triggered:
            self.active_grains = []
            self.last_buffer = np.zeros(frames)
            return out

        def get_curved_env(phase, curve):
            if phase <= 0: return 0.0
            if phase >= 1: return 1.0
            if abs(curve) < 0.01: return phase
            if curve > 0:
                c = curve * 5.0
                return (np.exp(c * phase) - 1.0) / (np.exp(c) - 1.0)
            else:
                c = -curve * 5.0
                return 1.0 - (np.exp(c * (1.0 - phase)) - 1.0) / (np.exp(c) - 1.0)
                
        self.curved_env = 0.0
        if self.master_env > 0.001:
            if self.master_env >= 0.999 or self.is_triggered:
                self.curved_env = get_curved_env(self.master_env, self.params["atk_curve"][0])
            else:
                self.curved_env = 1.0 - get_curved_env(1.0 - self.master_env, -self.params["rel_curve"][0])

        def get_p(key):
            p = self.params[key]
            val = p[0]
            if p[5]:
                mod = self.get_mod_val(p[5])
                val += mod * p[6] * (p[2] - p[1])
            return max(p[1], min(p[2], val))

        fm_audio_signal = np.zeros(frames)
        rm_multipliers = np.ones(frames)

        if self.all_engines and self.MOD_ROUTES:
            for i in [1, 2]:
                route_idx = int(self.params[f"mod{i}_src"][0])
                if route_idx == 0: continue
                
                route = self.MOD_ROUTES[route_idx]
                amt = self.params[f"mod{i}_amt"][0]
                mod_eng = self.all_engines[route["src"]]
                
                mod_data = mod_eng.last_buffer
                if len(mod_data) != frames:
                    mod_data = np.zeros(frames)
                
                if route["type"] == "FM":
                    fm_audio_signal += mod_data * amt * 5.0
                elif route["type"] == "RM":
                    current_route_rm = (1.0 - amt) + (mod_data * amt * 5.0)
                    rm_multipliers *= current_route_rm

        if self.is_triggered:
            self.playback_pos = (self.playback_pos + get_p("scan") * step) % 1.0
        self.next_grain_time -= frames
        if self.next_grain_time <= 0 and len(self.active_grains) < int(get_p("max_grains")):
            self.next_grain_time = max(10, (SAMPLE_RATE / get_p("dens")))
            p_idx = (get_p("pos") + self.playback_pos) % 1.0
            
            base_pitch = get_p("pitch")
            p_mode = int(get_p("pitch_mode"))
            if p_mode == 1: # Semi
                base_pitch = 2.0 ** (base_pitch / 12.0)
            elif p_mode == 2: # Oct
                base_pitch = 2.0 ** base_pitch
            
            new_g = Grain(
                self.sample_data, 
                int(p_idx * len(self.sample_data)), 
                get_p("size") * SAMPLE_RATE, 
                base_pitch, 
                0.5 + (random.random() - 0.5) * get_p("spread"), 
                random.random() < get_p("direction"), 
                self.window_lut[self.SHAPES[int(get_p("shape"))]], 
                get_p("grain_amp")
            )
            self.active_grains.append(new_g)
        
        mixed = np.zeros((frames, 2))
        for g in self.active_grains[:]:
            mixed += g.process(frames, fm_audio_signal)
            if not g.active: self.active_grains.remove(g)
            
        # Ring Modulation (RM) block
        if self.all_engines and self.MOD_ROUTES:
            for i in [1, 2]:
                route_idx = int(self.params[f"mod{i}_src"][0])
                if route_idx == 0: continue
                
                route = self.MOD_ROUTES[route_idx]
                if route["type"] == "RM":
                    amt = self.params[f"mod{i}_amt"][0]
                    mod_data = self.all_engines[route["src"]].last_buffer
                    
                    # Mix Dry/Wet based on amt (scale x5.0 matching C++ engine)
                    rm_mix = (1.0 - amt) + (mod_data * amt * 5.0)
                    mixed[:, 0] = mixed[:, 0] * rm_mix
                    mixed[:, 1] = mixed[:, 1] * rm_mix
        
        f_type = int(get_p("filt_type"))
        if f_type != 3:
            cut, res = get_p("cutoff"), get_p("res")
            for i in range(frames):
                mixed[i, 0] = self.filter_l.process(mixed[i, 0], cut, res, f_type)
                mixed[i, 1] = self.filter_r.process(mixed[i, 1], cut, res, f_type)
        
        final_signal = (mixed * (self.params["vol"][0] * self.curved_env))
        self.last_buffer = np.clip((final_signal[:, 0] + final_signal[:, 1]) * 0.5, -1, 1)
            
        return final_signal.astype(np.float32)